Oracle.md

```diff
@@ -0,0 +1,162 @@
+# Oracle specifications
+
+## Introduction
+
+For the purpose of these specifications, an event is a digital representation of a real world fact.
+An [oracle](./Introduction.md#Oracle) is an entity that commits to publishing a signature over a (number of) event outcome(s) ahead of time by releasing an [R-value](./Introduction.md#R-value) as well as necessary information for two parties to build a set of [CETs](./Introduction.md#Contract-Execution-Transaction-(CET)).
+This necessary information is committed in a so-called _event descriptor_ that will be further detailed in this document.
+Descriptors are classified and discovered through a namespace presently represented as Unique Resource Identifier (URI) paths.
+
+## Event descriptor
+
+An event descriptor provides information to clients about an event for which an oracle plans on releasing a signature over its outcome.
+The provided information should be sufficient for a client to create a set of CETs that will cover all possible outcomes of the event.
+
+Here we assume that an event outcome can be represented either as a set of strings or numbers.
+Four kinds of event outcomes are defined (note that a single even can be composed of several types).
+
+### External event descriptors
+
+External event descriptor enables representation of oracles in a format agreed upon between an oracle and its clients, for the cases where the following descriptor types are inadequate.
+The structure and semantic of the information contained in the descriptor is thus open.
+
+### Simple enumeration
+
+For events that have a narrow range of possible outcomes, the outcomes can simply be enumerated.
+For the sake of simplicity, outcomes should be represented as strings.
+
+#### Example: Weather tomorrow
+
+Tomorrow's weather can be represented as the set of strings `[sunny, cloudy, rainy]`.
+
+### Range
+
+When an event has numerical outcomes that cannot be easily enumerated, they can be represented as a series with:
+
+- start: the first possible outcome number
+- count: the number of possible outcomes
+- step: the increment
+
+#### Example: tomorrow's temperature
+
+```
+start: -100
+count: 201
+step: 1
+```
+
+### Large range
+
+When a range of a numerical outcomes is large or unbounded, the oracle can represent the outcome using numerical decomposition.
+
+The event descriptor should include:
+
+- base: the base in which the outcome value is decomposed and which will indicate the possible range of each digit that will be signed.
+- is-signed: a boolean value indicating whether the outcomes can be negative.
+- r-values: an array of R values, one for each of the digit that will be signed by the oracle. If the is-signed value is true, an additional R-value must be provided as the first element in this array, which will be used to sign the string "+" in case of an outcome with a positive value and the string "-" in case of an outcome with a negative one.
```

Comment:
Could you add more clarity here, it is unclear which nonces refer to which digit

ie if signing up to `999` does r-values[0] refer to the 1s place or the 100s place

Reply:
Imo it should refer to the 1s place, that way you don't need to know the size at all times to know which rvalue refers to which digit

Reply:
I was thinking about the other way, because you need to use the R value for the 100s more than you do the one for the 1s in average. Also I'm not sure I understand the point about needing to know the size at all time, maybe could you provide an example of what the issue is, or how it makes things more convenient the way you are proposing?

(I'll definitely clarify once we have an agreement on which way to go)

Reply:
Thinking more what I said above is probably not true, you actually do use the one for the 1s more than the one for the 100s. And I think I actually understand your point. But thinking that I added the sign R value as first element, maybe it kind of breaks the thing anyway, or would you put the R value for sign at the end?

Reply:
Yeah I think you'd want the lowest place first in the array and then have the last be the sign

Reply:
So actually I agree with Tibo's first position that the first R should be the sign and then the highest order digit down to the one's place. When it comes to the _signatures_ which is what you should be concerned with when it comes to R values, the higher order digits' signatures are more important and later ones can often be ignored.

Reply:
> When it comes to the signatures which is what you should be concerned with when it comes to R values, the higher order digits' signatures are more important and later ones can often be ignored.

The oracle will always be signing for all of them so that shouldn't matter

Reply:
After implementing this it is actually pretty trivial to do either way so I am indifferent

Reply:
I mean I suppose this is not too important so long as it is standard (and I agree it makes no difference to the oracle) but for the clients it is slightly more convenient to have the biggest digit (most CET-relevant information) have the first digit